### PR TITLE
Refactor python interface & support pip

### PR DESCRIPTION
**Pull request by mzmssg**
```
# Install from github
pip install git+https://github.com/microsoft/nnfusion.git@[BRANCH]

# Install from github (without git tool)
pip install https://github.com/microsoft/nnfusion/archive/refs/heads/[BRANCH].zip

# Install from local
pip install /path/to/nnfusion/

# Install from github (develop mode)
pip install -e git+https://github.com/microsoft/nnfusion.git@[BRANCH]#egg=nnfusion
pip install -U -e git+https://github.com/microsoft/nnfusion.git@[BRANCH]#egg=nnfusion

# Install from local (develop mode)
pip install -e /path/to/nnfusion/

# Uninstall
pip uninstall nnfusion -y

# Test
python -c "import nnfusion;print(nnfusion.__version__)"
```

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    install_requires = f.read().splitlines()
+
+with open("VERSION") as f:
+    version = f.readline().strip()
+
+setuptools.setup(
+    name="nnfusion",
+    version=version,
+    author="Microsoft Corporation",
+    author_email="nnfusion@microsoft.com",
```
Comment (xysmlx) [inline setup.py:30]: Shall we use the external team mail alias "nnfusion-team" here?

File: src/python/example/mnist.py [hunk 2]
```diff
@@ -3,22 +3,25 @@
...
-from nnf.description import IODescription, generate_sample
-from nnf.trainer import Trainer
-import data_loader
 import json
-
-os.environ["PATH"] = os.path.abspath(
-    "./build/src/tools/nnfusion") + ":" + os.environ["PATH"]
+import tempfile
+
+from nnf.data_format import cast_numpy_array, cast_pytorch_tensor
```
Comment (xysmlx) [inline src/python/example/mnist.py:17]: I think it would be better to use nnfusion as the package name instead of nnf.

File: setup.py (hunk 1 above)
Comment (mzmssg) [reply setup.py:30]: Sure

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+numpy
+torch
+torchvision
```
Comment (xysmlx) [inline requirements.txt:3]: Do we need to add torchvision to requirements? It seems to be only used in the examples.

Comment (mzmssg) [reply requirements.txt:3]: Yes, it's for convenience. Another way is keeping a "requirements_test.txt" for all deps, then "requirements.txt" keeps only core deps. Then users might need to run additional `pip install torchvision == some version` before run example. Your opinion?

Comment (xysmlx) [reply requirements.txt:3]: I think it is good for convenience in current status.

**Comment by xysmlx** [comment]
It seems the pip package works when the nnfusion repo has been successfully built before the pip install. Do we need to support automatically building and installing nnfusion cli when installing the pip package?

BTW, shall we automatically import nnfusion cli into PATH after installing the pip package like the [nnictl](https://nni.readthedocs.io/en/latest/Tutorial/Nnictl.html) in NNI? Here is an [example](https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html).

**Comment by mzmssg** [comment]
> Do we need to support automatically building and installing nnfusion cli when installing the pip package.

Intalling nnfusion kit from some package tool is ultimate goal. But some platforms needn't (or not available) compile model, they only need runtime to execute compiled model. Besides, PIP is not a good tool for cmake project, though setuptools support simple cpp extension as well as build customization. Most cmake projects still need to maintain different prebuild distributions. APT might be better for this purpose.

>BTW, shall we automatically import nnfusion cli into PATH after installing the pip package like the nnictl in NNI? Here is an example.

It should be easy if we decide to include NNFusion binary in PIP distribution.

**Comment by mzmssg** [comment]
@xysmlx Actually this package is more like a runtime, we might use `nnfusion_runtime` as package name instead?

**Comment by xysmlx** [comment]
> @xysmlx Actually this package is more like a runtime, we might use `nnfusion_runtime` as package name?

I prefer 'nnfusion' as the package name as our ultimate goal is installing nnfusion from package tool. Currently, the way of installing nnfusion and python interface is like TVM. So I think we could still use nnfusion as the package name.

**Comment by mzmssg** [comment]
> BTW, how to activate the tests_require? I have tried `python setup.py test` and `python setup.py .[test]` but they did not work.

Tests_require won't be installed. Use pip install instead.

**Review by xysmlx — APPROVED** [review]

**Pull request merged**